play.pokemonshowdown.com/src/client-connection.ts

```diff
@@ -13,70 +13,173 @@ declare const POKEMON_SHOWDOWN_TESTCLIENT_KEY: string | undefined;
 export class PSConnection {
 	socket: any = null;
 	connected = false;
-	queue = [] as string[];
+	queue: string[] = [];
+	private reconnectDelay = 1000;
+	private reconnectCap = 15000;
+	private shouldReconnect = true;
+	private worker: Worker | null = null;
+
 	constructor() {
 		const loading = PSStorage.init();
 		if (loading) {
 			loading.then(() => {
-				this.connect();
+				this.initConnection();
 			});
 		} else {
+			this.initConnection();
+		}
+	}
+
+	initConnection() {
+		const workerConnected = this.tryConnectInWorker();
+		if (!workerConnected) {
 			this.connect();
 		}
 	}
+
+	tryConnectInWorker(): boolean {
+		try {
+			const worker = new Worker('/js/reconnect-worker.js');
+			this.worker = worker;
+
+			worker.postMessage({ type: 'connect', server: PS.server });
+
+			worker.onmessage = event => {
+				const { type, data } = event.data;
+				switch (type) {
+				case 'connected':
+					console.log('\u2705 (CONNECTED via worker)');
+					this.connected = true;
+					PS.connected = true;
+					this.queue.forEach(msg => worker.postMessage({ type: 'send', data: msg }));
+					this.queue = [];
+					PS.update();
+					break;
+				case 'message':
+					PS.receive(data);
+					break;
+				case 'disconnected':
+					this.handleDisconnect();
+					if (this.shouldReconnect) this.retryConnection();
+					break;
+				case 'error':
+					console.warn('Worker connection error');
+					this.worker = null;
+					this.connect(); // fallback
+					break;
+				}
+			};
+
+			worker.onerror = (e: ErrorEvent) => {
+				console.warn('Worker connection error:', e);
+				this.worker = null;
+				this.connect(); // fallback
+			};
+
+			return true;
+		} catch {
+			console.warn('Worker connection failed, falling back to regular connection.');
+			this.worker = null;
+			return false;
+		}
+	}
+
 	connect() {
+		this.worker = null; // ensure worker isn't used. we can keep
```

Comment:
This doesn't actually stop the worker from being used, though. I think `if (this.worker) return` is safer.